packages/swayze/lib/src/widgets/headers/gestures/header_gesture_detector.dart

```diff
@@ -250,67 +273,231 @@ class _HeaderGestureDetectorState extends State<HeaderGestureDetector> {
     );
   }
 
+  /// Handles drag starts that should start dragging a header around.
+  void handleStartDraggingHeader(
+    DragStartDetails details,
+    Range selectionRange,
+  ) {
+    setCursorState(SystemMouseCursors.basic);
+    Actions.invoke(
+      context,
+      HeaderDragStartIntent(
+        draggingPosition: details.localPosition,
+        headers: selectionRange,
+        axis: widget.axis,
+      ),
+    );
+  }
+
+  /// Handles header dragging updates.
+  void handleUpdateDraggingHeader(
+    DragUpdateDetails gestureDetails,
+    _HeaderGestureDetails details,
+  ) {
+    Actions.invoke(
+      context,
+      HeaderDragUpdateIntent(
+        draggingPosition: gestureDetails.localPosition,
+        header: details.headerPosition,
+        axis: widget.axis,
+      ),
+    );
+  }
+
+  void handleDragEnd(SwayzeHeaderDragState state) {
+    if (state.isDropAllowed) {
+      Actions.invoke(
+        context,
+        HeaderDragEndIntent(
+          header: state.dropAtIndex,
+          axis: widget.axis,
+        ),
+      );
+    } else {
+      handleDragCancel();
+    }
+  }
+
+  void handleDragCancel() {
+    Actions.invoke(
+      context,
+      HeaderDragCancelIntent(widget.axis),
+    );
+  }
+
+  /// Sets a new cursor state.
+  void setCursorState(MouseCursor newCursor) {
+    if (newCursor != cursor) {
+      setState(() => cursor = newCursor);
+    }
+  }
+
+  /// Checks if a header is selected.
+  bool isHeaderSelected(int position, Axis axis) =>
+      internalScope.controller.selection.isHeaderSelected(position, axis);
+
+  /// Finds the selection that is under position.
+  ///
+  /// Returns null if position header is not selected.
+  HeaderUserSelectionModel? hoverSelection(int position, Axis axis) {
+    final selectionController = internalScope.controller.selection;
+    final selections = selectionController.userSelectionState.selections;
+
+    for (final selection in selections) {
+      if (selection is HeaderUserSelectionModel && selection.axis == axis) {
+        final range = Range(selection.start, selection.end);
+        if (range.contains(position)) {
+          return selection;
+        }
+      }
+    }
+    return null;
+  }
+
+  /// Finds the range of the current reference selection and all adjacent
+  /// selections.
+  ///
+  /// Returns a range containing all adjacent selections from the reference
+  /// selection, so it can be dragged as a single group.
+  Range headerSelectionRange(HeaderUserSelectionModel referenceSelection) {
+    final selectionController = internalScope.controller.selection;
+    final selections = selectionController.userSelectionState.selections;
+    var selectionRange =
+        Range(referenceSelection.start, referenceSelection.end);
+    for (final selection in selections) {
```

Comment:
If we're iterating over selections in their order, isn't there a chance that not all adjacent selections are selected?

Let's say we have the following selections, in order:

```dart
final selections = [Range(start: 1, end: 2), Range(start: 2, end: 3)];
```

This is similar to having 3 separate selections of the columns A, B and C (or equivalent rows), right?

In this case, if `referenceSelection` is `Range(start: 0, end: 1)`, the returned `selectionRange` is `Range(start: 0, end: 3)`, right?


However, if the selections are, in order:

```dart
final selections = [Range(start: 2, end: 3), Range(start: 1, end: 2)];
```

If `referenceSelection` is `Range(start: 0, end: 1)`, the returned `selectionRange` will be `Range(start: 0, end: 2)`, right? Is that intended?

Reply:
good catch